R/autoqmd_dependency_stamp.R

```diff
@@ -0,0 +1,147 @@
+#' Generate dependency timestamp from a QMD
+#'
+#' Parse a QMD (via `knitr::purl`), detect `source()` calls (including
+#' `file.path(...)` forms), attempt to evaluate simple assignments that appear
+#' before the first `source()` (so variables become available),
+#' and collect `mtimes` of the file and any dependency files.
+#'
+#' @param file Path to a `.qmd` file.
+#' @param quiet Logical; if TRUE suppresses messages about unresolved sources.
+#'
+#' @return A single string: an HTML comment with dependency `mtimes`, e.g.
+#'   `<!-- DEPENDENCY-MTIME: template.qmd 2025-10-08 14:30:11 -->`
+#'
+#' @seealso [autoqmd_insert_includes()]
+#'
+#' @export
+#'
+#' @importFrom knitr purl
+#' @importFrom rprojroot find_root_file
+#'
+#' @examples
+#' \dontrun{
+#' autoqmd_dependency_stamp("species_template.qmd")
+#' }
+autoqmd_dependency_stamp <- function(file = NULL, quiet = FALSE) { # nolint: cyclocomp_linter
+  # If no valid file -> fallback to current time
+  if (is.null(file) || !file.exists(file)) {
+    return(sprintf("<!-- DEPENDENCY-MTIME: %s -->",
+                   format(Sys.time(), "%Y-%m-%d %H:%M:%S")))
+  }
+
+  # 1) Extract R code from the file (works for QMD / knitr chunks)
+  tmp_r <- tempfile(fileext = ".R")
+  knitr::purl(file, output = tmp_r, documentation = 0L, quiet = TRUE)
+  lines <- readLines(tmp_r, warn = FALSE)
+
+  # 2) Find indices of lines that contain source(
+  src_idx <- grep("source\\(", lines)
+  if (length(src_idx) == 0) {
+    # No sources found -> return file mtime only
+    mt <- file.info(file)$mtime
+    stamp <- paste0(basename(file), " ", format(mt, "%Y-%m-%d %H:%M:%S"))
+    return(sprintf("<!-- DEPENDENCY-MTIME: %s -->", stamp))
+  }
+
+  # 3) Prepare a restricted environment for safe evaluation
+  env <- new.env(parent = baseenv())
+  # provide helpers that are commonly used in files
+  env$file.path <- file.path
+  env$here <- function(...) file.path(...) # minimal here()-like helper
+
+  # If rprojroot is available, provide the function directly
+  if (requireNamespace("rprojroot", quietly = TRUE)) {
```

Comment:
We can require `rprojroot`. This removes the `if` statement and reduces the cyclomatic complexity

```suggestion
  stopifnot("`rprojroot`is not installed" = requireNamespace("rprojroot", quietly = TRUE))
```